bson-codec-design/DESIGN.md

```diff
@@ -0,0 +1,177 @@
+# BSON Encoder & Decoder Design
+The v1 design for the BSON library's encoder and decoder contain design issues that
+require a redesign. The main design flaw is the lack of a coherent, flexible design that can
+incorporate the needs of users. Assumptions made during the v1 design have proven incorrect or too
+inflexible to be as useful as initially intended.
+
+## Definitions
+<dl>
+<dt>Codec Processing</dt>
+<dd>The complete process of either encoding or decoding.</dd>
+<dt>Unmarshal</dt>
+<dd>The process of converting a slice of bytes into a native Go type</dd>
+<dt>Marshal</dt>
+<dd>The process of converting a native Go type into a slice of bytes</dd>
+<dt>Encode</dt>
+<dd>The process of converting a native Go type into BSON and writing it to an
+<code>io.Writer</code></dd>
+<dt>Decode</dt>
+<dd>The process of reading a BSON document from an <code>io.Reader</code> and converting it to a
+native Go type</dd>
+<dt>v1 Design </dt>
+<dd>The initial design for the encoding and decoding functionality of the BSON library</dd>
+<dt>Proposed Design</dt>
+<dd>The design laid out in this document</dd>
+</dl>
+
+## Motivation
+The v1 design for codec processing in the BSON library made a number of assumptions about the
+control the user had over their types, the amount of granular control users would want to have over
+codec processing, and the flexibility required during codec processing. A key initial assumption of
+the v1 design was that users who needed to handle types in a custom way would be able to make those
+types implement the `bson.Marshaler` and `bson.Unmarshaler` interfaces. This turned out to be too
+restrictive because users sometimes want to handle the processing for types they do not define
+themselves. For example, users who generate types using a protocol buffer library
+or users who want to handle the processing of types in the standard library.
+
+The design of the v1 encoder and decoder for the Go driver presents problems when users wish to
+have more control over how specific types are handled when encoding or decoding. The control that
+users require exists on two main levels:
+- control over how individual types are encoded or decoded
+- control over how tags are processed
+
+The granularity of control users require can be defined in three levels:
+- Control over how individual types are encoded and decoded when they are embedded in another type,
+such as a map or a struct
+- Control over how individual types are encoded and decoded when they are provided directly to the
+encoder or decoder
+- Control over how tags are processed and the parameters that are returned
+
+The proposed design allows for the first two granularities to be handled together.
+
+## High Level Design
+### Registries
```

Comment:
Somewhere in this document, state where a Registry can be configured for use by the mongo package, e.g.

* MongoClient creation/configuration
* Database/Collection creation/configuration
* Individual CRUD methods

I'd recommend at least the first, possibly the second, and not the third

Reply:
This will go in a separate design doc.